Add support for path exploration to CBMC/JBMC, in addition to full bounded model checking

karkhaz opened the pull request:
CBMC can now model-check each path of a program one at a time, displaying the verification result for that path before continuing to symbolically execute other paths. This functionality is similar to what the `symex` tool provides, but uses the `goto-symex` core that is already used by CBMC rather than having a disjoint symbolic execution codebase. This functionality is enabled by passing the `--paths` flag to CBMC or JBMC.

When `--paths` is passed, CBMC does not perform path merging when it encounters a program join point. Instead, whenever it encounters a conditional `goto` instruction, it saves the state of the symbolic executor onto a workqueue and executes one of the branches. Once symbolic execution has finished and the verification result printed out, CBMC pops the workqueue and resumes executing the untaken branch.

This is all integrated with the existing `bmc` and `goto_symext` classes, with no change in CBMC's original model-checking functionality. Without the `--paths` flag, CBMC's full-program model-checking behaviour is an edge-case of path exploration, whereby no goto branch-points are saved and so the entire program is only symbolically executed once.

The main patch in this patchset is `[path explore 5/6] Support path-based exploration`. Other commits are fairly small and introduce features that are necessary for path exploration, but which are also desirable changes by themselves.

karkhaz committed:
[path explore 2/7] Ignore jbmc binary

karkhaz committed:
[path explore 4/7] Factor out common BMC code

The CBMC and JBMC frontends included identical code for running BMC
after the language-specific preprocessing. This commit moves that common
code into a static method of bmct.

Parse options and help text related to bounded model checking are
defined in bmc.h.

martin-cs commented:
Thanks for this patch set; it looks really exciting.

Paging @theyoucheng who has a similar patch set and @bjowac who was the original author of symex.  Also cross linking against https://github.com/diffblue/cbmc/issues/749 for my last attempt at getting everyone who cared about per-path symex talking to each other so we didn't duplicate work.

karkhaz committed:
[path explore 5/7] Support path-based exploratio2

This commit introduces the --paths flag to CBMC, which makes CBMC
model-check each individual path of the program rather than merging
paths and model-checking the entire program.

The overall strategy involves allowing the "state" of symbolic
execution---i.e., a pair of goto_symex_statet and symex_target_equationt
---to be saved during an execution, and for an execution to be resumed
from a saved pair. By saving the state at every branch point and
disabling path merging, symbolic execution only runs along one path but
populates a worklist of saved states that should be executed later. At
the top level, CBMC or JBMC loops while the worklist of saved states is
non-empty, creating a new bmct object to resume executing each saved
path.

This commit includes the following supporting changes:

- goto_symex_statet now owns a symbol table, separate from the one
  supporting the goto_program, which is used for dynamically created
  objects during symbolic execution. goto_symex was previously using a
  symbol table that was passed to it by reference for this purpose, but
  that symbol table is needed when resuming symbolic execution from a
  saved point and so ought properly to be part of the symbolic execution
  state.

- goto_symex_statet now has a pointer to a symex_target_equationt, which
  can be updated with an equation from a previously-saved symbolic
  execution run. While equations are also conceptually part of the state
  of symbolic execution, they are heavily used after symbolic execution
  has completed (and the symex state has been deallocated) and so the
  equation is not owned by the state. An explicit copy constructor has
  been added to goto_symex_statet that initializes the equation member,
  so that symbolic execution can proceed either with an empty equation
  or with an equation that was earlier saved.

- goto_symex_statet no longer has a pointer to a dirtyt, as this was
  hindering it from being copied.

karkhaz commented:
@martin-cs thanks for the comment. This is a complete implementation and is working well, but I'd like to braindump a few future refinements here for the record:

- Different heuristics for deciding which path to execute next. The implementation in this pull request is a simple FIFO, i.e. whenever we get to a branch point, we save the path that we don't execute for now, and resume paths in the order that they were saved. It turns out that this isn't a great strategy. I should have better ones in the next few weeks.

- Multithreaded CBMC. This is exciting! Every path that is symbolically executed could be done in a different thread. The only data structure that needs to be shared is the workqueue where upcomming paths are saved, so we'd need to lock that (and it's potentially a fairly hot data structure). Mark Tuttle suggests that to remedy this, we could experiment with every thread having its own local queue of paths to execute next, and pull paths from the global workqueue only when the local one gets emptied (and of course, every thread should contribute paths to resume to the global workqueue).
Another way to avoid this issue entirely is to have all symbolic execution happen in the main thread, but fork off the SAT solver in a different thread or process (so that CBMC is symbolically executing one path while the SAT solver checks another).
In either case, thread/process control would be done entirely from the method `bmct::do_language_agnostic_bmc`, which is a method that is introduced in this patchset. I expect that doing this wouldn't be too invasive, but we'd need to add an implementation of a thread or process pool to the codebase.

cheers!

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
+    const optionst &opts,
+    const goto_modelt &goto_model,
+    const ui_message_handlert::uit &ui,
+    messaget *const message)
```
Take by reference?

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
+      result=bmc.run(goto_model.goto_functions);
+    }
+    INVARIANT(opts.get_bool_option("paths") || worklist.empty(),
+        "Some branches were unexplored when model-checking "
```
Recommend making a "should" statement here, as otherwise it's often confusing whether the statement says what should have been or what went wrong.

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
+      worklist.pop_front();
+    }
+  }
+  catch(const char *error_msg)
```
Recommend throwing structured errors rather than plain strings wherever possible

karkhaz replied:
@smowton this catch is to deal with a throw from `cbmc_solverst`, i.e. it's code that predates this patchset. As far as I can see, that code only ever throws the number `0` (and it prints a message to `messaget::error()` just before it throws). I'm not familiar with cprover's error handling, what sort of error would you recommend that I throw from `cbmc_solvert` instead of `0`? Or should I just leave that there?

smowton replied:
"Pretty crap," in brief ;) If it's old code you're catching from, I guess you intend to catch everything? In that case I'd catch `const char *` and `const std::string &` (perhaps mergable if exception catching permits implicit construction to catch, but I doubt it) and then `catch(...)` to catch other errors. If you specifically intend to catch *some* things but let others get by then you should at least document your intent in comments, and probably change the throwing code to throw something more specific.

martin-cs replied:
#include "martin_comment_about_catch_all_vs_invariants.h"

If you are going to catch (...) please exit and / or throw upwards; many exceptions are non-recoverable.

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
+      cbmc_solver=solvers.get_solver();
+      prop_convt &pc=cbmc_solver->prop_conv();
+      goto_symext::branch_pointt &resume=worklist.front();
+      path_explorert pe(opts,
```
If there are this many args, probably clearest to break after `(` and then list one argument per line

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
     symex.set_unwind_limit(options.get_unsigned_int_option("unwind"));
 }
+
+int bmct::do_language_agnostic_bmc(
```
Please add function-level doxy comments

karkhaz replied:
The prevalent style seems to be having the doxy comments in header files, right? I already have one there, but it's a bit brief so I'll expand it.

smowton commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -845,6 +798,8 @@ void jbmc_parse_optionst::help()
     " --stop-on-fail               stop analysis once a failed property is detected\n" // NOLINT(*)
     " --trace                      give a counterexample trace for failed properties\n" //NOLINT(*)
+    " --paths                      explore paths one at a time (implies\n"
+    "                              '--goto-priority jump')\n"
```
...if `--goto-priority` is not manually specified

smowton commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -110,6 +108,16 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
+  if(cmdline.isset("goto-priority"))
+    options.set_option("goto-priority", cmdline.get_value("goto-priority"));
+  else if(options.get_bool_option("paths"))
+    options.set_option("goto-priority", "jump");
```
Maybe use an enum for this in `optionst`?

karkhaz replied:
@smowton I don't understand. How would I be able to call `optionst::set_option()` with an enum as an argument? The only way I can think of would be to have a dedicated `optionst::set_goto_priority()` function and have the goto_priority be a field inside `optionst`, is this what you meant? Or am I misunderstanding?

smowton replied:
No, I meant just using `optionst`'s `set_option(unsigned int)` overload and corresponding `get_unsigned_int_option` with casts to and from an enum type. It's not a lot better since `optiont` will stringify and then `atoi` it internally and doesn't enforce type, but at least a `switch` on some `enum goto_priorityt` will yield incomplete switch-case warnings from your compiler if and when it grows a third option.

smowton commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -110,6 +108,16 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
+  if(cmdline.isset("paths"))
+    options.set_option("paths", true);
+
+  if(cmdline.isset("goto-priority"))
```
Mentioned this below, but make clear that even with `--paths` this can be manually overridden.

smowton commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -56,6 +57,7 @@ class optionst;
   "(error-label):(verbosity):(no-library)" \
   "(nondet-static)" \
   "(version)" \
+  "(paths)(goto-priority):" \
```
Since these are duplicated between cbmc and jbmc and probably other driver programs will want them, suggest moving the arg specs and help text to an appropriate header file and including them here, as is already done for `OPT_GOTO_CHECK` for example.

smowton commented on src/goto-instrument/accelerate/scratch_program.cpp:
```diff
@@ -39,7 +39,8 @@ bool scratch_programt::check_sat(bool do_slice)
 
-  symex(symex_state, functions, *this);
+  symex.symex_with_state(symex_state, functions, *this, new_symbol_table);
+  ns=namespacet(symbol_table, new_symbol_table);
```
Suggest this should be initialised in `scratch_programt`'s constructor, since both `symbol_table` and `new_symbol_table` are (presumably `new_symbol_table` is harmlessly empty until this has happened in any case?)

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,49 @@ class safety_checkert:public messaget
+    safety_checkert::resultt const &r1,
+    safety_checkert::resultt const &r2)
+{
+  return r1==safety_checkert::resultt::SAFE?safety_checkert::resultt::SAFE:
```
Spaces around `?` and `:`, and `=` below (some of the CBMC codebase has compressed spacing, but thankfully conventional spacing is now allowed :))

smowton commented on src/goto-symex/goto_symex.h:
```diff
@@ -63,22 +86,27 @@ class goto_symext
+    symex_target_equationt *const saved_equation,
+    symbol_tablet &new_symbol_table);
 
   /** start symex in a given state */
```
Good opportunity to improve this comment? For example, if we start from here, where do we continue until?

smowton commented on src/goto-symex/goto_symex.h:
```diff
@@ -128,20 +157,20 @@ class goto_symext
-  const namespacet &ns;
-  symex_targett &target;
+  const symbol_tablet &outer_symbol_table;
+  namespacet ns;
```
Add comment indicating why we have a mutable one rather than a const ref as previously? Does this always point to `outer_symbol_table`, or...?

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -126,12 +146,62 @@ void goto_symext::symex_goto(statet &state)
+    goto_programt::const_targett tmp=new_state_pc;
+    new_state_pc=state_pc;
+    state_pc=tmp;
+    std::cerr << "Resuming from '" << state_pc->code.source_location() << "'\n";
```
If this is a debug message, guard with `#ifdef DEBUG_GOTO_SYMEX` or something (and the same for `iostream` include above). Otherwise, use a `messaget` or get one passed in.

karkhaz replied:
If #1625 gets merged in before this, symext will have a messaget mutable member as you suggested so I'll rebase and use that

tautschnig replied:
FWIW I'll merge #1625 as soon as CI completes.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -99,6 +100,25 @@ void goto_symext::symex_goto(statet &state)
+  do_simplify(simpl_state_guard);
+
+  // No point executing both branches of an unconditional goto.
+  if(new_guard.is_true()&&    // We have an unconditional goto, AND
```
Space before `&&` and `||`

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -126,12 +146,62 @@ void goto_symext::symex_goto(statet &state)
+  // Goto is conditional, so figure out which branch to take before
+  // checking whether to do path exploration or model-checking, and before
+  // adjusting guards.
+  std::string gp=options.get_option("goto-priority");
```
As above, suggest using an enum type

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -126,12 +146,62 @@ void goto_symext::symex_goto(statet &state)
+    branch_point.state.has_saved_target=true;
+    branch_point.state.saved_target_is_backwards=forward;
+    branch_worklist.push_back(branch_point);
+    std::cerr << "Saving '" << new_state_pc->source_location << "'\n";
```
Another use of std streams

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -159,58 +161,87 @@ void goto_symext::symex_threaded_step(
+  const goto_programt &goto_program,
+  symbol_tablet &new_symbol_table)
 {
+  ns=namespacet(outer_symbol_table, state.symbol_table);
```
As above, could you comment why namespaces are being re-initialised to point at different tables like this? What tables will `ns` refer to and at what stage in proceedings?

smowton reviewed (commented):
Some comments. Mostly doxy comments, style and some inconsistency in how the arguments are described. Most substantively I'd appreciate comments where appropriate about why namespaces are being re-initialised to point at different symbol tables, and at what stage they point to different places (therefore, if a future coder is editing `goto_symext` and adds an `ns.lookup`, what tables do they think they're referring to? Is it outright invalid (uninitialised) at any point?)

smowton commented:
Also edited title to the usual "voice" for such things :)

karkhaz commented:
@smowton thanks so much for the prompt review, I'll try and get this done within the next few days.

martin-cs commented:
@karkhaz Multi-threading certainly is exciting but may I add a small note of caution.  SAT solvers are very sensitive to memory latency and cache sizes.  Given parts of the memory subsystem are shared between cores, running a SAT solver in one core can adversely affect the performance of a SAT solver running in another.  Before you start work, I'd recommend trying the following experiment.

*. Pick a benchmark that takes at least 30 seconds and uses at least at least 50 MB of memory in the SAT solver.
*. Try running 1, 2, ..., $NUMBER_OF_CORES copies simultaneously and record the *wall clock* time taken to finish the complete run.
*. Plot total time and total time / N (work-rate) against N and see where they peak.

If total time is flat and work-rate peaks at N; brilliant, my warning no longer applies.  If the peak is elsewhere and the lower it is, consider if it is worth the investment of time to multi-thread stuff.

( Also, a seeded pseudo-random queuing policy, run as a portfolio could also give a lower bound / estimate on the performance of the multi-threading.)

The other exciting topic, with @DanielNeville can provide input on is when to merge and when to queue.  I'm told that "merge forwards branches and queue backwards ones" does interesting things...

karkhaz commented:
@martin-cs thanks for this information! I'd certainly be interested in better merging heuristics. Something that I have working (not in this patchset) is the ability to put `__CPROVER_begin_path_explore()` and `__CPROVER_end_path_explore()` dummy function calls in the target codebase. The idea is to be able to do merging up until a point, and then do path exploration from that point onward, and switch merging back on. Or, the other way round. This very coarse-grained approach seems to work nicely for particular "shapes" of codebases. But obviously an _automatic_ decision on when to merge would be extremely welcome.

Hopefully there's a lot of experimenting to be done here: what are the best heuristics for deciding when to merge, and what are the best heuristics for deciding on which path to explore next. I'm planning to gather some amount of empirical data about this, and would be very interested to read data gathered by others also.

karkhaz committed:
[path explore 6/7] cpplint & clang-fmt agree on :

cpplint now complains if a derived-class declaration is formatted as

  class derived: base

rather than complaining if it is formatted as

  class derived : base

(with a space on both sides of the colon). The latter style is the one
enforced by clang-format, meaning that the two linters were inconsistent.

karkhaz commented:
@smowton I've addressed all your suggestions, thanks so much!

I've also added two commits to deal with the cpplint and clang-format. They're in this PR because linting won't pass without them.

smowton commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -18,53 +18,52 @@ Author: Daniel Kroening, kroening@kroening.com
-  "(localize-faults)(localize-faults-method):" \
+#define CBMC_OPTIONS                                                           \
+  OPT_BMC                                                                      \
+  "(preprocess)(slice-by-trace):" OPT_FUNCTIONS                                \
```
Keep `OPT_FUNCTIONS` and `OPT_GOTO_CHECK` on their own lines

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,59 @@ class safety_checkert:public messaget
+  safety_checkert::resultt const &r1,
+  safety_checkert::resultt const &r2)
+{
+  return r1 == safety_checkert::resultt::SAFE
```
How about defining an `operator<` then simply returning `std::min` or `std::max(r1, r2)`?

smowton replied:
(or a named comparator function, say `result_worse_than`, then using `std::min(r1, r2, result_worse_than)`

smowton replied:
Just realised this is only showing up at all due to clang-format, so can ignore this

smowton commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -829,88 +772,97 @@ void jbmc_parse_optionst::help()
-    " --json-ui                    use JSON-formatted output\n"
-    " --verbosity #                verbosity level\n"
-    "\n";
+  std::cout << "* *              Daniel Kroening, Edmund Clarke             * "
```
I'd avoid reformatting the whole help text / arg list if I were you, even if the linters complain -- this makes an annoying commit to rebase across

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -369,9 +372,8 @@ safety_checkert::resultt bmct::execute(const goto_functionst &goto_functions)
-  statistics() << "size of program expression: "
-               << equation.SSA_steps.size()
-               << " steps" << eom;
+    statistics() << "size of program expression: " << equation.SSA_steps.size()
```
Don't reformat unchanged lines

martin-cs replied:
+1 what he said.

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -595,3 +597,143 @@ void bmct::setup_unwind()
+  if(cmdline.isset("goto-priority"))
+  {
+    const std::string &priority = cmdline.get_value("goto-priority");
+    if(priority == "jump")
```
Braces around multi-line `if` (similarly below)

smowton commented on src/cbmc/bmc.h:
```diff
@@ -26,24 +28,53 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "symex_bmc.h"
 
-class bmct:public safety_checkert
+class goto_modelt;
```
Header is included so probably unnecessary

smowton reviewed (commented):
Looks good! Couple of nitpicks; I'd probably drop commit #8 though (the wholesale linter patch), clang-format is introducing lots of pointless churn here. Though perhaps I'm doing a King Canute act -- @peterschrammel any opinion?

tautschnig commented:
Echoing @smowton's comment: please don't `clang-format` what hasn't been touched to avoid rebasing hell.

martin-cs commented:
@tautschnig and @smowton I've run into issues with clang format's incremental use not being that incremental.

karkhaz commented:
@smowton that's done. I removed the forward class declaration. All your other comments were about changes introduced in the clang-format commit, which I've dropped.

smowton commented:
Great, thanks! In that case my only other suggestion is to replace the exhaustive listing of `resultt` with a comparator and `min` and `max` calls in `operator||` and `&&` respectively. Otherwise all looks great!

karkhaz commented:
@smowton cheers, I did that. I think this implementation is actually better, as it uses a switch statement so that the compiler can warn if new values are added to the enum, rather than checking and throwing at runtime.

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,39 @@ class safety_checkert:public messaget
 
+/// \brief is `r1` a worse result than `r2`?
+inline bool operator<(
+    safety_checkert::resultt &r1,
```
Take by value

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,39 @@ class safety_checkert:public messaget
+  switch(r1)
+  {
+  case safety_checkert::resultt::ERROR:
+    return true;
```
This is wrong, two things that are `==` must not be `<`. This should return true if r1 is *strictly worse than* r2.

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,39 @@ class safety_checkert:public messaget
+}
+
+/// \brief The worst of two results
+inline safety_checkert::resultt &operator&=(
```
Take and return by value

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,39 @@ class safety_checkert:public messaget
+}
+
+/// \brief The best of two results
+inline safety_checkert::resultt &operator|=(
```
Similar

smowton reviewed (commented):
Like the pattern, couple of problems with the implementation

karkhaz commented:
@smowton I did ask you asked, but note that the first argument to an assignment operator must be passed by non-const reference, not by value (otherwise the assignment cannot happen). So I'm passing in the second argument by value, and returning by value.

smowton commented:
Gah, sorry, the change to `operator<` is correct; `operator&=` and `operator|=` should be references as you had them previously. I'd misread them as `operator&` and `operator|` respectively.

Then this lgtm. Really this time. Honest.

karkhaz commented:
@smowton No worries :) that's done.

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,37 @@ class safety_checkert:public messaget
+  case safety_checkert::resultt::SAFE:
+    return false;
+  case safety_checkert::resultt::ERROR:
+    return r2==safety_checkert::resultt::ERROR ? false : true;
```
`return r2 != safety_checkert::resultt::ERROR`

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,37 @@ class safety_checkert:public messaget
+  case safety_checkert::resultt::ERROR:
+    return r2==safety_checkert::resultt::ERROR ? false : true;
+  case safety_checkert::resultt::UNSAFE:
+    return r2==safety_checkert::resultt::SAFE ? true: false;
```
return `r2 == safety_checkert::resultt::SAFE`

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,37 @@ class safety_checkert:public messaget
+    safety_checkert::resultt &a,
+    safety_checkert::resultt const &b)
+{
+  a = std::min((safety_checkert::resultt)a, (safety_checkert::resultt)b);
```
No need for these casts anymore

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -45,4 +45,37 @@ class safety_checkert:public messaget
+    safety_checkert::resultt &a,
+    safety_checkert::resultt const &b)
+{
+  a = std::max((safety_checkert::resultt)a, (safety_checkert::resultt)b);
```
Nor these

smowton reviewed (approved):
Nitpicks, no need for re-review from me. Thanks for making the effort, and apologies again for my mistake :)

karkhaz commented:
thanks so much! done with nitpicks. should I ping anybody in particular to merge this in? @tautschnig is likely busy this week.

smowton commented:
Suggest @martin-cs to give a final sign off on the design principles?

martin-cs commented on .gitignore:
```diff
@@ -99,6 +99,7 @@ src/goto-cc/goto-cc.exe
 src/goto-instrument/goto-instrument
 src/goto-instrument/goto-instrument.exe
+src/jbmc/jbmc
 src/musketeer/musketeer
```
Yes!

martin-cs commented on scripts/cpplint.py:
```diff
@@ -3625,8 +3625,8 @@ def CheckOperatorSpacing(filename, clean_lines, linenum, error):
-    error(filename, linenum, 'readability/identifier_spacing', 4, 'There shouldn\'t be a space between class identifier and :')
+  if Search(r'(struct|class)\s[\w_]*:', line):
+    error(filename, linenum, 'readability/identifier_spacing', 4, 'There should be a space between class identifier and :')
 
```
*sigh* Another lint vs. clang-format issue.  Thanks for fixing.

martin-cs commented on src/cbmc/bmc.h:
```diff
@@ -126,6 +137,35 @@ class bmct:public safety_checkert
+  " --partial-loops              permit paths with partial loops\n" \
+  " --no-pretty-names            do not simplify identifiers\n" \
+  " --graphml-witness filename   write the witness in GraphML format to filename\n" // NOLINT(*)
 };
```
It's good that these become common options.

martin-cs commented on src/analyses/dirty.h:
```diff
@@ -17,42 +17,74 @@ Date: March 2013
-  explicit dirtyt(const goto_functiont &goto_function)
+  dirtyt():
+    initialized(false)
+  {
```
It's no exactly clear why this is needed.  I guess there must be some reason why you have to create and initialise it separately.

martin-cs commented on src/analyses/dirty.h:
```diff
@@ -17,42 +17,74 @@ Date: March 2013
 
+  void build(const goto_functionst &goto_functions)
+  {
+    INVARIANT(!initialized,
```
I think this should be a PRECONDITION because it's a condition on the structure which must hold for this to be called,

martin-cs commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -493,9 +499,9 @@ void goto_symex_statet::rename(
   {
-    assert(expr.operands().size()==1);
+    PRECONDITION(expr.operands().size()==1);
     rename_address(expr.op0(), ns, level);
```
I'd say DATA_INVARIANT because this is something that should always be true of a well formed address-of

martin-cs commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -493,9 +499,9 @@ void goto_symex_statet::rename(
     rename_address(expr.op0(), ns, level);
-    assert(expr.type().id()==ID_pointer);
+    PRECONDITION(expr.type().id()==ID_pointer);
     expr.type().subtype()=expr.op0().type();
```
Ditto.

martin-cs commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -512,7 +518,7 @@ void goto_symex_statet::rename(
     {
-      assert(to_if_expr(expr).true_case().type()==
+      PRECONDITION(to_if_expr(expr).true_case().type()==
              to_if_expr(expr).false_case().type());
```
Ditto.

martin-cs commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -99,6 +99,25 @@ void goto_symext::symex_goto(statet &state)
+    // we're going to skip over it for now and return to it later.
+     options.get_bool_option("paths")))
+  {
+    INVARIANT(instruction.targets.size()>0,
```
DATA_INVARIANT : this should always be true.

martin-cs commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -102,9 +103,12 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
     usage_error();
-    exit(1); // should contemplate EX_USAGE from sysexits.h
+    exit(CPROVER_EXIT_USAGE_ERROR);
   }
```
Yes.

martin-cs commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -203,7 +207,7 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
             << "must not be given together" << eom;
-    exit(1); // should contemplate EX_USAGE from sysexits.h
+    exit(CPROVER_EXIT_USAGE_ERROR);
   }
```
Yes.

martin-cs commented on src/cbmc/bmc.h:
```diff
@@ -287,7 +299,10 @@ class path_explorert:public bmct
+  " --goto-priority <jump|next>  jump: execute the target of a goto first.\n" \
+  "                              next: execute the instruction following\n" \
+  "                              the goto first. Default: 'next', unless\n" \
+  "                              --paths is passed.\n" \
```
I think it would be better to have a more general --path-strategy or --path-selection-heuristic or something similar because this gives two strategies but it would be nice to have the option able to support others.  Depth-first, unexplored-first, etc.

martin-cs reviewed (commented):
Most of it looks good.  5/7 is a big patch and I don't quite see the architecture of the new part, can you describe it a bit more?  I think 6/7 should be a bit more general to support future options.

thk123 commented on src/analyses/dirty.h:
```diff
@@ -17,42 +17,74 @@ Date: March 2013
-    forall_goto_functions(it, goto_functions)
-      build(it->second);
+    build(goto_functions);
+    // build(g_funs) responsible for setting initialized to true, since
```
This possibly isn't your design - but it would seem to be better for this constructor only to be visible, then you don't need to separately track initialised and have constructors as the only way build this object (of course removing the default constructor). (Possibly mirroring Martin's comment above).

karkhaz replied:
Thanks thk123. I believe that this isn't possible:
- `goto_symex_statet` has a `dirtyt` member, which needs to be initialized upon creation of the `goto_symex_statet`.
- However, `goto_symex_statet` is not actually supplied with a goto_functions object until just before it is used for symbolic execution---it does not receive a goto_functions object when it is constructed. The code was previously designed so that `statet`s were only constructed when needed in `goto_symex_main` since they take so much memory, and so they are not a member of `goto_symext`.
- Therefore, we need to initialize the dirtyt only after the statet has been given a goto_functionst.

Does this make sense? I'll add comments clarifying this. Note that the code had previously been doing a similar (but less safe) thing: statet had a _pointer_ to a dirtyt instead of a member, and the dirtyt was a `nullptr` until supplied with a goto_functions. I changed this because we now are actually copying states around and copying objects that contain stale pointers turned out to lead to much more obscure error messages.

thk123 replied:
OK, the only way round this I can think of is having a interface `dirtyt`, an `unitialized_dirtyt` and an `initalized_dirtyt` that inherit from it. Then the unitialized one could just have `INVARIANT(false)`) on all the relevant method calls and the `initlaized_dirtyt` can only be constructed with the right values. 

However, not even convinced myself this is better so I won't block the PR on it.

thk123 reviewed (commented):
Just reviewed the `src/analyses` and `scripts/` stuff. I think it would be much better use constructors to enforce correct construction of `dirtyt` so you don't need run time checks.

thk123 reviewed (approved):
Reviewed `scripts` and `analyses` changes

martin-cs commented:
@karkhaz (How) is this going?  I'm still very enthusiastic about the concept!

karkhaz commented:
@martin-cs I've been bogged down with various things. I'll try to finish up the changes and push them up by Friday!

martin-cs commented:
@karkhaz At your convenience :-)  I probably won't have much review time after the next hour or so until mid next week.

karkhaz commented:
@martin-cs @smowton I'm afraid this is going to have to wait another week, I'm bogged down. Should have an update late next week!

martin-cs commented:
I've waited a number of years for someone to do this properly; I'm sure
that a week or two more won't hurt ;-)

karkhaz committed:
[path explore 7/7] Path exploration documentation

karkhaz commented:
@martin-cs @smowton applied all your suggestions, thanks! Since I kept you waiting for so long, I wrote an additional documentation patch. As well as briefly describing what BMC is doing in [5/8], I've also added some higher-level architectural information in [8/8] on the symbolic execution module documentation page.

The SSA stuff and other ramblings are based on a conversation I had with @kroening, but any mistakes in that text are mine. Pinging @tautschnig in case that documentation needs a look-over.

cheers!

karkhaz commented:
An update on the rebase: @smowton introduced some code commented with `// unwinds <clinit> loops to number of enum elements` in commit `06a220ace5` in `jbmc_parse_options.cpp` which modifies the `bmct` that `jbmc_parse_optionst` calls `do_bmc()` on. Since my pull requests delegates the construction of the `bmct` object to a static method of `bmct`, I've moved @smowton's code into a static method in `bmct` called `unwind_clinit_loops()`. I'm doing this because my path-exploration code constructs multiple bmct objects, and I presumably need to call `add_loop_unwind_handler()` on all of them.

Please let me know if that seems sensible, the new code is at `bmc.h:135` and called from `bmc.cpp:624` and `bmc.cpp:670`.

karkhaz committed:
[path explore 3/7] Logical ops for resultt

The operators &= and |= can now be used between two
safety_checkert::resultt objects. They return the "best" or "worst" of
two results, and additionally assign that result to the LHS.  The
assignment operators are useful when repeatedly performing a safety
check and wanting to return a single error code that either describes
the best or worst outcome of all the safety checks combined.

smowton commented:
@karkhaz that's not quite sane because the code you've added it to is language-agnostic, but the feature is highly Java-specific. I suggest you should allow the frontend to supply an optional (functiont) hook, call it "driver_program_configure_bmc" or something, and call it each time you make a bmct. Then cbmc will currently supply no hook, while jbmc will supply one that registers its loop handler.

smowton commented:
@karkhaz @tautschnig I've created https://github.com/diffblue/cbmc/pull/1844 which is just my `dirtyt` and symex changes, which are hopefully pretty minor. As far as I can see, to rebase on top of them all you'd need to do would be `s/goto_functionst/get_goto_functiont/` and `s/dirtyt/incremenetal_dirtyt`.

Suggested sequence of operations:

1. Merge https://github.com/diffblue/cbmc/pull/1844
2. Rebase and merge this
3. I'll rebase https://github.com/diffblue/cbmc/pull/1759

Sound okay?

karkhaz commented:
@smowton SGTM, I'm currently working on your bmc configure-hook suggestion. I'll rebase as soon as #1844 is merged, thanks!

karkhaz commented:
@smowton I've added the hook function, please let me know if it's sensible. It's added in [4/8]. The hook function is called from `bmc.cpp` [here](https://github.com/diffblue/cbmc/pull/1641/files#diff-6b42a28511965b646db79583668c5e6eR668) and [here](https://github.com/diffblue/cbmc/pull/1641/files#diff-6b42a28511965b646db79583668c5e6eR623); the declaration, default implementation, and a brief usage comment is in [`bmc.h`](https://github.com/diffblue/cbmc/pull/1641/files#diff-966e54fda2e5f1e251ddac83557cb641R98); and [`jbmc_parse_options.cpp`](https://github.com/diffblue/cbmc/pull/1641/files#diff-a5e185f91169176ecd28d83be54fe404R495) supplies a custom implementation if `--java-unwind-enum-static` has been passed.

smowton commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -491,55 +492,34 @@ int jbmc_parse_optionst::doit()
+      ui_message_handler.get_ui(),
+      *this,
+      [](bmct &bmc, const goto_modelt &goto_model) { // NOLINT (*)
+        bmc.add_loop_unwind_handler([&goto_model](
```
Break after `(`, i.e.
```
bmc.add_loop_unwind_handler(
  [&goto_model](
    const irep_idt &function_id,
    unsigned loop_number,
    unsigned unwind,
    ...
```

smowton commented:
Also suggest instead of repeating the whole bmct call, which invites accidental deviation in the future, do something like
```
std::function<...> configure_bmc;
if(option_set)
  configure_bmc = ...
else
  configure_bmc = noop
bmc.language_agnostic(...)
```

Other than these small style nitpicks this looks good to me :)

karkhaz commented:
@smowton I think this is done, thanks!

smowton commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -491,55 +492,31 @@ int jbmc_parse_optionst::doit()
+        };
+  return bmct::do_language_agnostic_bmc(
+    options, goto_model, ui_message_handler.get_ui(), *this, configure_bmc);
+  UNREACHABLE;
```
No need for this

smowton reviewed (commented):
No need for `return; UNREACHABLE;`, otherwise lgtm

karkhaz commented:
OK, done. Cheers

smowton commented:
1844 has landed, please rebase!

karkhaz commented:
@smowton I won't be able to complete this today, I need some time to understand the changes; the rebase isn't entirely trivial. Hope to have this done over the weekend!

smowton commented:
No worries, thanks for your patience and hard work here! A quick English summary of what was done in 1844:
* Symex switched from using a `goto_functions_mapt` to a function `get_goto_function`, of type `irep_idt -> goto_functiont &`. As of now the only user implements that by `goto_functionst::at`.
* Symex's entry-points that use `goto_functionst` are simple proxies to the `get_goto_function` version of the same method.
* `dirtyt` gains a method to incrementally analyse methods, and a wrapper class `incremental_dirtyt` that analyses functions that first time it sees each one.

martin-cs commented:
@karkhaz : Likewise, thank you for your patience.  Let me know when you've had a chance to look at this and I can re-review.

karkhaz commented:
@smowton @martin-cs I've rebased, but the build keeps timing out, not sure what to do about this. I forgot to diff the rebase against my original patch, but the changes I made are:

- `goto_symex_statet` has an `incremental_dirtyt` member, rather than `unique_ptr<dirtyt>` (original) or `unique_ptr<incremental_dirtyt>` (as of 1844) or `dirtyt` (as of my original PR);

- I re-introduced the `initialized` member to `dirtyt`. The `dirtyt::add_function()` function introduced in 1844 now sets `initialized` to true, without checking whether it was already set, since I suppose that this function might properly be called several times whenever some new java bytecode is loaded.

- Added a convenience `populate_dirty_for_functions` to complement the `populate_dirty_for_function`

- `goto_symex_main` I _think_ I left all the 1844 code intact, but re-introduced the `new_symbol_table` parameter to many of the functions.

- In `goto-symex/symex_function_call.cpp` I removed a check for whether the `dirty` is a null pointer since it's no longer a pointer

tautschnig commented:
I have restarted the Travis job, hoping it was just one of the many random failures.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -126,12 +148,71 @@ void goto_symext::symex_goto(statet &state)
+    branch_pointt branch_point(target, state);
+    branch_point.state.saved_target = new_state_pc;
+    branch_point.state.has_saved_target = true;
+    branch_point.state.saved_target_is_backwards = forward;
```
I think this is correct, but add a comment clarifying the counterintuitive "backward = forward" statement here

karkhaz commented:
Thanks @tautschnig, that worked! FWIW I tried to "restart" the job by `git commit --amend`ing and re-pushing, but it timed out again. Anyway, ready for review. Cheers!

smowton commented:
I don't see that you've broken anything re 1844-- two nonblocking queries:

1. Are there any places where you use goto_functionst  and need more than the functionality of get_goto_functions? If so, what do you need? If not, could you make a get-goto-function from the functions map as soon as possible, as is done for symex-from-entry-point for example?

2. Is there any need for a populate-dirty-all-at-once operation? It seems to me if we always use it incrementally this will both save time in the case that symex reveals large amounts of code to be unreachable, and also means there is only one uniform code path to maintain.

smowton commented:
I'm away on holiday this week but will be checking email and eager to get this in so I can rebase the rest of my work on top of it :)

karkhaz commented:
Not sure if I understood query 1, does the change that I just pushed resolve it? I added code to `resume_symex_from_saved_state` similar to what you have in `symex_from_entry_point_of`.

I've just removed the populate all at once function. It was being called only from a constructor that I added to `goto_symex_statet` (`explicit goto_symex_statet::goto_symex_statet(goto_functionst &)`) which I never used. Not sure why I added that, but it's now gone, too. Hope this helps!

smowton commented:
Yes, that looks good -- I may add a public entry point that takes a get-goto-function directly, but I'm happy for that to be my problem.

One more query that occurred to me while reviewing: are the various new virtual functions virtual for a reason, or just anticipating some potential future subclass? If the latter then in the cbmc codebase even public functions are usually non virtual until a second implementation comes along.

karkhaz commented:
Are you referring to the functions in `goto_symext`? The only totally new function that I introduced is `resume_symex_from_saved_state`, and indeed it's virtual like the others...I made it virtual because I assumed that somebody else was overriding them, but now I see the only subclass is `symex_bmct` and it doesn't override any of those functions. Should I remove the `virtual` keyword from all of the `symex_` functions in a separate commit and avoid introducing it in my code in [5/8]?

tautschnig commented on src/cbmc/bmc.h:
```diff
@@ -54,39 +91,106 @@ class bmct:public safety_checkert
-  virtual ~bmct() { }
+  virtual ~bmct()
+  {
+  }
```
Shouldn't need reformatting?

tautschnig commented on src/cbmc/bmc.h:
```diff
@@ -54,39 +91,106 @@ class bmct:public safety_checkert
-  // the safety_checkert interface
-  virtual resultt operator()(
-    const goto_functionst &goto_functions)
+  void set_ui(ui_message_handlert::uit _ui)
```
Looks like a reformatting change?

tautschnig commented on src/cbmc/bmc.h:
```diff
@@ -54,39 +91,106 @@ class bmct:public safety_checkert
-  void add_unwind_recursion_handler(
-    symex_bmct::recursion_unwind_handlert handler)
+  // the safety_checkert interface
+  virtual resultt operator()(const goto_functionst &goto_functions)
```
Reformatting?

tautschnig commented on src/goto-programs/safety_checker.h:
```diff
@@ -14,6 +14,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
 // this is just an interface -- it won't actually do any checking!
 
+#include <algorithm>
```
What is that header needed for in this file?

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -177,36 +214,48 @@ class goto_symext
   // c) rewrite array_equal expression into equality
-  void clean_expr(
-    exprt &, statet &, bool write);
+  void clean_expr(exprt &, statet &, bool write);
```
Unnecessary reformatting

tautschnig commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -22,8 +20,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
-symex_target_equationt::symex_target_equationt(
-  const namespacet &_ns):ns(_ns)
+symex_target_equationt::symex_target_equationt()
```
How about removing this explicit definition and instead use `= default;` with the declaration?

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
+    "* * Carnegie Mellon University, Computer Science Department * "
+    "*\n"
+    "* *                 kroening@kroening.com                   * "
+    "*\n"
```
I'd suggest not to reformat.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
+    "property is detected\n" // NOLINT(*)
+    " --trace                      give a counterexample trace for "
+    "failed properties\n" //NOLINT(*)
+    "\n" HELP_FUNCTIONS
```
I think this is just reformatting? Please don't.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
     "\n"
-    "Program instrumentation options:\n"
-    HELP_GOTO_CHECK
+    "Program instrumentation options:\n" HELP_GOTO_CHECK
```
No reformatting, please.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
+    " --cover CC                   create test-suite with coverage "
+    "criterion CC\n" // NOLINT(*)
+    " --mm MM                      memory consistency model for "
+    "concurrent programs\n" // NOLINT(*)
```
No reformatting.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
+    // This one is handled by jbmc_parse_options not by the
+    // Java frontend, hence its presence here:
+    " --java-unwind-enum-static    try to unwind loops in static "
+    "initialization of enums\n" // NOLINT(*)
```
Don't reformat.

tautschnig commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -878,87 +829,94 @@ void jbmc_parse_optionst::help()
-    HELP_TIMESTAMP
+    " --json-ui                    use JSON-formatted "
+    "output\n" HELP_GOTO_TRACE
+    " --verbosity #                verbosity level\n" HELP_TIMESTAMP
```
Please review all these help-text changes for unnecessary reformatting.

tautschnig commented on src/jbmc/jbmc_parse_options.h:
```diff
@@ -31,21 +33,22 @@ class optionst;
   "(object-bits):" \
   "(classpath):(cp):(main-class):" \
-  "(depth):(partial-loops)(no-unwinding-assertions)(unwinding-assertions)" \
+  "(partial-loops)" \
```
`partial-loops` goes together with all the unwinding options, could you please move it to `OPT_BMC` as well?

tautschnig reviewed (approved):
There's a fair bit of reformatting that should be please removed from the PR. Otherwise it looks good.

karkhaz commented:
Thanks @tautschnig! I've removed all formatting changes, the needless inclusion of `<algorithm>` in `safety_checker.h`, and also written a comment clarifying the `backwards = forwards` assignment in `goto-symex/symex_goto.cpp` as @smowton suggested.

tautschnig commented:
@martin-cs @peterschrammel @chrisr-diffblue Would you mind taking a look? @kroening any concerns?

chrisr-diffblue reviewed (approved):
Looks good to me - thanks to all the other good reviews and Kareem's diligent work.

karkhaz commented:
Please don't merge just yet---I may have found a problem! Will update very shortly!

karkhaz commented:
OK, let's wait for the tests to go through and then it should be fine to merge. Everything is looking good on my end. Sorry for the last-minute panic!

smowton commented:
Virtual functions: fair enough if they are normal for this class. Thanks again, all looks good to me!

martin-cs commented on src/cbmc/bmc.h:
```diff
@@ -82,6 +83,25 @@ class bmct:public safety_checkert
+    std::function<void(bmct &, const goto_modelt &)> frontend_configure_bmc =
+      [](bmct &_bmc, const goto_modelt &) { // NOLINT (*)
+        // Empty default implementation
+      });
```
This feels a bit clunky and not really in keeping with how the rest of the code base does extensions of functionality.  Is there a simpler way of doing this?  Also the apparent use-case for this (loop unwinding for the Java front-end) may be going away / getting transformed, maybe @mgudemann or @smowton might be able to say more.

peterschrammel replied:
I think this should be cleaned up in the course of https://github.com/diffblue/cbmc/issues/1869

martin-cs commented on src/goto-symex/goto_symex.h:
```diff
@@ -34,28 +35,49 @@ class symbol_exprt;
       atomic_section_counter(0),
       log(mh),
-      guard_identifier("goto_symex::\\guard")
+      guard_identifier("goto_symex::\\guard"),
```
Should this be guard coded?  I realise it was before...

karkhaz replied:
I don't understand the question, sorry: do you mean hard-coded? If so, what would be the alternative, is there some kind of symbolic constant for this?

martin-cs commented on src/goto-symex/symex_target_equation.h:
```diff
@@ -291,14 +292,12 @@ class symex_target_equationt:public symex_targett
   }
 
-  void output(std::ostream &out) const;
-
```
Why are you removing the output method?

karkhaz replied:
Because I removed the `ns` member, which is needed for that function. This was all part of my effort to clarify what classes own what objects, but I'm afraid I don't recall exactly why `ns` was a member of this class (perhaps @tautschnig could elaborate? I think we discussed this and I forgot...)

When I wanted to create the `branch_pointt` class (consisting of an equation and a state), I tried to remove all members that would point to objects that might become stale, since we need to resume the equation and state later and stale/out-of-scope references led to segfaults. I _think_ the `ns` member was one of the victims of this purge, and so was the `dirtyt` member (that's why I changed it from a pointer to a member)...

tautschnig replied:
I think we weren't able to figure out why `ns` should possibly be a member here. Indeed @martin-cs is right that `output` needs review. We should have a method that is explicitly passed a `namespacet` as an argument. That implies that the `operator<<` declarations should go away as well (their definitions have been removed!). I'd suggest to re-instantiate the `output` method exactly as it was defined before, except that it should take an additional `namespacet` parameter.

karkhaz replied:
OK, will push that in a few minutes

martin-cs reviewed (commented):
[2/8] YES.
[3/8] Sure.
[4/8] Yes; we should have done this when JBMC was first forked.
[5/8] Yes, MUCH better with explanations.
[6/8] Maybe?  It seems like this is adding names for two particular search heuristics, "depth-first-take-goto" and "depth-first-ignore-goto".  To be honest I'd rather see a more general option / infrastructure that can easily be extended or otherwise I fear we will see a profusion of options like this that do similar-but-not-the-same things.
[7/8] Thanks for fixing.
[8/8] YES.

martin-cs commented:
I realise this has been going on for some while and fear we are trying @karkhaz 's considerable patience ... BUT ... can I say "accept everything apart from 6/8"?

My objection to 6/8 isn't ideological, if it gets accepted, it gets accepted. I just think it will get changed pretty soon and we could easily get it right now and save hassle later.  Symbolic execution heuristics (just "which path next, not even 'should I merge' or 'when should I feasibility check') are (somehow still?) an active research topic (@DanielNeville @cesaro @theyoucheng and others may be able to comment).  It's one of the obvious low-hanging fruit with this.  Many of these will require non-orthogonal changes to the branch taken behaviour "ignore forwards branches, follow backwards branches", "follow backwards branches but only the first 2 times", "take everything that looks like an error path because no-one checks those", "pick at random because I want to pretend I'm fuzzier than I am" which feel kind of hard to square with the proposed 6/8 solution.  So what I would /suggest/ (I'm not going to demand) is that you swap it out for a more general "heuristic" option that not only picks which way to go but also sorts the work queue.  The two options would then be depth-first-follow-goto and depth-first-ignore-goto or something.

karkhaz commented:
@martin-cs there's two separate "sites" where we chose what to do next. The first site is deciding which path to put onto the workqueue (this is what is controlled by the command line option in 6/8), and the second site is deciding what to take off the workqueue (which is what I believe you're talking about).

You're totally correct that the choice of _what to take off the workqueue_ will change soon, and can be decided by advanced heuristics. I'm working on this now, and that code will go into `bmct::do_language_agnostic_bmc()` (since this is where we resume executing new paths, so this is where we should decide which path to resume). However, at the point where we're _saving_ new paths, there are literally only ever going to be two choices, because `goto` instructions have a one target instruction and one next instruction. The `--goto-priority` flag is about choosing between the goto's target and the goto's next instruction to push onto the workqueue, and this will remain the case even when we come up with cool heuristics for deciding what to pop off the workqueue. Does this make sense?

I wonder whether you're advocating for heuristics to control _both_ pushing onto the queue and popping off? I suppose I could try that as well, if that's what you mean...

martin-cs commented:
@karkhaz From an implementation point of view, yes it makes complete sense.  However I think it might be good to think about things a little more theoretically.  Per-path symex generates a (likely infinite) tree.  Any heuristic (including parallel exploration!) is then an ordering of nodes in the tree (i.e. an ordering with the tree as a sub-order).  In implementation terms, it's like you add both new states to the work queue, and then pick the next state from the work queue.  If you're working depth-first then it will be one of the two (well... assuming a step-based notion of depth) but maybe you want to be able to explore breadth first or some kind of beam search.

I'm not going to block things over this but it seems like any of the more advanced heuristics, even depth-first ones, will want to control both of the locations you describe, so having a flag that fixes the behaviour at one seems a bit limiting.

tautschnig commented:
@karkhaz Just state whether you'd like to review the code and make changes right away in light of @martin-cs' comments.

@martin-cs & @karkhaz: I'd like to merge this relatively soon to have a first implementation in place for everyone to use and then iterate to improve it.

karkhaz commented:
@tautschnig sorry, I pushed right before I saw your comment. @martin-cs I've removed 6/8, and I'll add a more general flag once I have my heuristics ready and it's more clear what the implementation will look like. Does that work?

martin-cs commented:
@tautschnig I think my question about what's happened to the output method is probably more of a blocker than the 6/8 question.

martin-cs commented:
@karkhaz Thank you that is most considerate of you.

karkhaz commented:
Just pushed: removed the formatting fix, and reinstated the `output` method with a new `namespacet` argument. Not sure what to do about the hard-coded `guard` identifier, and I'm not sure how I would have implemented the bmc customization differently; it does appear to need some kind of callback as it currently stands. If the use-case is going away then we can simply omit that callback parameter from `bmct::do_language_agnostic_bmc()` when that happens, it shouldn't be a difficult change, but why not keep it for now---I think the avoidance of code duplication that `do_language_agnostic_bmc` gives us more than makes up for any custom callbacks that we need to pass into that function.

karkhaz commented:
Also, going to sleep now, I have a flight tomorrow that I need to wake up for...should have time to make any additional changes tomorrow early morning if needed.

tautschnig commented:
@karkhaz Thanks a lot for all those updates! I'm inclined to merge once CI passes, unless @martin-cs tells me otherwise?

martin-cs commented:
> I don't understand the question, sorry: do you mean hard-coded?

Yes; sorry.

>  If so, what would be the alternative, is there some kind of symbolic constant for this?

Not yet but have a look at util/cprover_prefix.h for the current
direction of travel of this sort of thing.

This one isn't urgent or blocking as it's not your change.

martin-cs reviewed (approved):
@tautschnig Nah; I think we're good here.

tautschnig commented:
CI has passed successfully, just need @martin-cs to give the green light.

karkhaz commented:
Thanks so much everyone for the detailed reviews and suggestions! Looking forward to continuing work on this.

peterschrammel commented on src/cbmc/bmc.h:
```diff
@@ -28,24 +31,57 @@ Author: Daniel Kroening, kroening@kroening.com
     message_handlert &_message_handler,
-    prop_convt &_prop_conv)
+    prop_convt &_prop_conv,
+    goto_symext::branch_worklistt &_branch_worklist)
```
For more backwards compatibility, would it have been possible to provide a constructor that initialises branch_worklist to a default value here?

tautschnig replied:
I'm not sure default values and references would be compatible?

smowton commented:
@karkhaz just coming to rebase my own work on top of this; a couple of questions:

1. At the moment this looks like it will interact oddly with any of the reporting options (show-vcc, program-only, symex-coverage-report) by outputting a report each time bmc is run. Is that intentional, or should the iterative symex-then-bmc run until "done", then only run end-of-day reporting once?

2. How should this interact with the (default) all-properties mode or with `--cover`, both of which try to cover multiple goals? At present it is only merging a single safe / not-safe return value, which feels as if it's designed for use with `--stop-on-fail`, and with `--cover` it would presumably repeatedly print reports saying that (for a particular symex path) goals X, Y and Z were reached, then for another path goals A, B and C were... and so on. Should we declare --paths currently only for  use with --stop-on-fail, or is there pending work to merge the goal map to produce a single coherent report later?

karkhaz commented:
@smowton: not sure if I have a good answer for these questions, I'm not actually that familiar with many of CBMC's command line options.

1. I'm not sure why anybody would pass `--paths` together with reporting options like `--program-only`? If you want to do `--program-only`, you presumably want to symex the whole program at once without doing any checks, rather than doing one path at a time, right? Perhaps I'm not understanding the use-case here, but it seems to me that you would only want to pass `--paths` if you're actually interested in checking properties.

2. I'm not sure what the "goal map" is. I wasn't using `--paths` with `--stop-on-fail`, I was indeed reading the diagnostic output from the BMC result one at a time while the loop was spinning. Admittedly this is not at all elegant, and it would be great to have some sort of unified report at the end that collects the BMC results of all paths seen so far. So again I'm not totally sure if I understand the question, but I would not be opposed to making `--paths` imply `--stop-on-fail` for now until I come up with a cleaner way of presenting the results from all paths. I suppose that I'd trap `SIGINT` or whatever and print a report before exiting.

smowton commented:
Re: #1, it would potentially be useful if you wanted to see the equation before each run, which is indeed what would currently happen, but in such a case we should somehow prepend each report with a header identifying a path.

Re #2, the three options 'cover', 'fault-localization' and the default all-properties mode all build a list of all the assertions they're trying to satisfy, then make repeated solver runs to try to satisfy different goals. I guess with --paths we should probably stop trying to target a particular goal once it has been found using a particular path? So the goal list would be maintained outside the twin iterations over goals and paths, and goals incrementally reported either hit or covered as we go, before finally reporting ones we couldn't reach? Or am I getting the wrong end of the stick, and it's actually interesting to check if I can violate goal A via both paths X and Y?

martin-cs commented:
( Unsolicited opinions on what the options should do)

--paths --show-vcc            prints each formula instead of running the solver
--paths --show-program    prints each path condition instead of running the solver
--paths --all-properties      tries to prove every property using the per-path approach
--paths --cover                  tries to cover each location using the per-path approach
--paths --fault-localization ask David Landsberg or @theyoucheng
